controllers/disruption_controller.go

```diff
@@ -192,53 +193,73 @@ func (r *DisruptionReconciler) Reconcile(ctx context.Context, req ctrl.Request)
+				}()
+
+				return ctrl.Result{}, nil
+			} else if instance.Status.CleanedAt == nil {
+				isCleaned, err := r.cleanDisruption(ctx, instance)
+				if err != nil {
+					return ctrl.Result{}, fmt.Errorf("error cleaning disruption: %w", err)
+				}
 
-			// we reach this code when all the cleanup pods have succeeded
-			// we can remove the finalizer and let the resource being garbage collected
-			r.log.Infow("all chaos pods are cleaned up; removing disruption finalizer")
-			r.recordEventOnDisruption(instance, chaosv1beta1.EventDisruptionFinished, "", "")
+				// if not cleaned yet, requeue and reconcile again in 15s-20s
+				// the reason why we don't rely on the exponential backoff here is that it retries too fast at the beginning
+				if !isCleaned {
+					requeueAfter := time.Duration(randSource.Intn(5)+15) * time.Second //nolint:gosec
 
-			r.DisruptionsWatchersManager.RemoveAllWatchers(instance)
-			controllerutil.RemoveFinalizer(instance, chaostypes.DisruptionFinalizer)
+					r.log.Infow(fmt.Sprintf("disruption has not been fully cleaned yet, re-queuing in %v", requeueAfter))
 
-			if err := r.Client.Update(ctx, instance); err != nil {
-				return ctrl.Result{}, fmt.Errorf("error removing disruption finalizer: %w", err)
-			}
+					return ctrl.Result{
+						Requeue:      true,
+						RequeueAfter: requeueAfter,
+					}, r.Client.Update(ctx, instance)
+				}
 
-			// send reconciling duration metric
-			r.handleMetricSinkError(r.MetricsSink.MetricCleanupDuration(time.Since(instance.ObjectMeta.DeletionTimestamp.Time), []string{"disruptionName:" + instance.Name, "namespace:" + instance.Namespace}))
-			r.handleMetricSinkError(r.MetricsSink.MetricDisruptionCompletedDuration(time.Since(instance.ObjectMeta.CreationTimestamp.Time), []string{"disruptionName:" + instance.Name, "namespace:" + instance.Namespace}))
-			r.emitKindCountMetrics(instance)
+				// set cleanedAt
+				now := metav1.Now()
+				instance.Status.CleanedAt = &now
 
-			// close the ongoing disruption tracing Span
-			defer func() {
-				_, disruptionStopSpan := otel.Tracer("").Start(ctx, "disruption deletion", trace.WithAttributes(
-					attribute.String("disruption_name", instance.Name),
-					attribute.String("disruption_namespace", instance.Namespace),
-					attribute.String("disruption_user", userInfo.Username),
-				))
+				if err := r.Client.Status().Update(ctx, instance); err != nil {
+					return ctrl.Result{}, fmt.Errorf("error updating disruption's status: %w", err)
+				}
 
-				disruptionStopSpan.End()
-			}()
+				r.recordEventOnDisruption(instance, chaosv1beta1.EventDisruptionCleaned, "", "")
+
+				requeueAfter := r.FinalizerDeletionDelay
+				r.log.Infow(fmt.Sprintf("all chaos pods are cleaned up; requeuing to remove finalizer after %s", requeueAfter))
+
+				return ctrl.Result{Requeue: true, RequeueAfter: requeueAfter}, nil
+			}
 
-			return ctrl.Result{}, nil
+			// requeue until we can remove the finalizer
+			return ctrl.Result{Requeue: true}, nil
```

Comment:
We should specify a requeueAfter, otherwise this will requeue nearly immediately